<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Purpose" data-toc-modified-id="Purpose-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Purpose</a></span></li><li><span><a href="#Import-libraries" data-toc-modified-id="Import-libraries-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Import libraries</a></span></li><li><span><a href="#Read-in-data" data-toc-modified-id="Read-in-data-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Read in data</a></span><ul class="toc-item"><li><span><a href="#Resting-heart-rate-data" data-toc-modified-id="Resting-heart-rate-data-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Resting heart rate data</a></span><ul class="toc-item"><li><span><a href="#Exploring-resting-heart-rate-data" data-toc-modified-id="Exploring-resting-heart-rate-data-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Exploring resting heart rate data</a></span></li><li><span><a href="#Extracting-resting-heart-rate-data" data-toc-modified-id="Extracting-resting-heart-rate-data-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Extracting resting heart rate data</a></span></li><li><span><a href="#Exporting-dailyHRdf-into-csv" data-toc-modified-id="Exporting-dailyHRdf-into-csv-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Exporting <code>dailyHRdf</code> into csv</a></span></li></ul></li><li><span><a href="#Sleep-data" data-toc-modified-id="Sleep-data-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Sleep data</a></span><ul class="toc-item"><li><span><a href="#Exploring-sleep-data" data-toc-modified-id="Exploring-sleep-data-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Exploring sleep data</a></span></li><li><span><a href="#Extracting-sleep-data" data-toc-modified-id="Extracting-sleep-data-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Extracting sleep data</a></span></li><li><span><a href="#Exporting-sleepdf-into-csv" data-toc-modified-id="Exporting-sleepdf-into-csv-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Exporting <code>sleepdf</code> into csv</a></span></li></ul></li><li><span><a href="#Step-data" data-toc-modified-id="Step-data-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Step data</a></span><ul class="toc-item"><li><span><a href="#Exploring-step-data" data-toc-modified-id="Exploring-step-data-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Exploring step data</a></span></li><li><span><a href="#Extracting-step-data" data-toc-modified-id="Extracting-step-data-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Extracting step data</a></span></li><li><span><a href="#Exporting-stepsdf-into-csv" data-toc-modified-id="Exporting-stepsdf-into-csv-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Exporting <code>stepsdf</code> into csv</a></span></li></ul></li><li><span><a href="#Activity-data" data-toc-modified-id="Activity-data-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Activity data</a></span><ul class="toc-item"><li><span><a href="#Exploring-activity-data" data-toc-modified-id="Exploring-activity-data-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Exploring activity data</a></span></li><li><span><a href="#Extacting-activity-data" data-toc-modified-id="Extacting-activity-data-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Extacting activity data</a></span></li><li><span><a href="#Exporting-actdf-data" data-toc-modified-id="Exporting-actdf-data-3.4.3"><span class="toc-item-num">3.4.3&nbsp;&nbsp;</span>Exporting <code>actdf</code> data</a></span></li></ul></li></ul></li><li><span><a href="#Summary" data-toc-modified-id="Summary-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Summary</a></span></li></ul></div>

### Purpose 
The purpose of this notebook is to read in all the extracted json files, put the required data into a DataFrame and then export them as csv files.

**Note:** There are 261 days worth of data extracted from `2021-02-25` to `2021-11-12`.

### Import libraries

In [10]:
from time import sleep
# import fitbit
import cherrypy
import requests

# import simplejson as json
import json
import datetime

import pandas as pd
import numpy as np

### Read in data

#### Resting heart rate data

##### Exploring resting heart rate data

In [12]:
# Loading in a first date
dailyHR = open('daily_HR2021-02-25.json')

# Loading json
dailyHRdata = json.load(dailyHR)
dailyHRdata = json.loads(dailyHRdata)

# Making it easier to read
print(json.dumps(dailyHRdata, indent=4, sort_keys=True))

{
    "activities-heart": [
        {
            "dateTime": "2021-02-25",
            "value": {
                "customHeartRateZones": [],
                "heartRateZones": [
                    {
                        "caloriesOut": 2334.132,
                        "max": 118,
                        "min": 30,
                        "minutes": 1427,
                        "name": "Out of Range"
                    },
                    {
                        "caloriesOut": 77.211,
                        "max": 145,
                        "min": 118,
                        "minutes": 13,
                        "name": "Fat Burn"
                    },
                    {
                        "caloriesOut": 0,
                        "max": 178,
                        "min": 145,
                        "minutes": 0,
                        "name": "Cardio"
                    },
                    {
                        "caloriesOut": 0,
                    

In [13]:
dailyHRdata.keys()

dict_keys(['activities-heart', 'activities-heart-intraday'])

In [14]:
dailyHRdata["activities-heart"][0].keys()

dict_keys(['dateTime', 'value'])

In [15]:
dailyHRdata["activities-heart"][0]["value"].keys()

dict_keys(['customHeartRateZones', 'heartRateZones', 'restingHeartRate'])

In [16]:
# format to extract resting heart rate
dailyHRdata["activities-heart"][0]["value"]["restingHeartRate"]

66

**Resting heart rate** seems to be nested within dictionary keys in `dailyHRData`.

##### Extracting resting heart rate data

In [17]:
dates = pd.date_range('2021-02-25', '2021-11-12')

restingHR = []
for date in dates:
    fname = 'daily_HR' + date.strftime('%Y-%m-%d') + '.json'
    with open(fname) as f:
        restingHR_data = json.loads(f.read())
        restingHRdata = pd.read_json(restingHR_data, typ='series')
        restingHR.append(restingHRdata["activities-heart"][0]["value"]["restingHeartRate"]) # the zero indexes the first value

In [18]:
# Check
restingHR

[66,
 67,
 66,
 66,
 65,
 65,
 66,
 67,
 66,
 66,
 64,
 65,
 66,
 69,
 69,
 69,
 67,
 67,
 68,
 66,
 65,
 65,
 64,
 65,
 65,
 66,
 65,
 64,
 64,
 64,
 64,
 64,
 63,
 65,
 65,
 65,
 64,
 66,
 68,
 65,
 66,
 65,
 63,
 64,
 63,
 62,
 61,
 62,
 63,
 64,
 64,
 64,
 63,
 65,
 65,
 64,
 67,
 66,
 67,
 65,
 65,
 67,
 66,
 63,
 63,
 63,
 64,
 62,
 63,
 63,
 63,
 63,
 64,
 64,
 64,
 65,
 65,
 65,
 65,
 67,
 66,
 66,
 65,
 65,
 64,
 65,
 66,
 65,
 66,
 65,
 66,
 65,
 64,
 64,
 64,
 64,
 63,
 63,
 64,
 64,
 66,
 66,
 67,
 66,
 66,
 68,
 69,
 67,
 66,
 67,
 65,
 65,
 65,
 65,
 63,
 64,
 66,
 65,
 65,
 66,
 67,
 67,
 67,
 66,
 65,
 64,
 65,
 65,
 65,
 66,
 65,
 64,
 64,
 64,
 65,
 65,
 65,
 67,
 65,
 64,
 65,
 65,
 65,
 66,
 65,
 64,
 65,
 64,
 63,
 63,
 63,
 65,
 65,
 64,
 63,
 63,
 64,
 64,
 65,
 64,
 64,
 64,
 66,
 67,
 65,
 66,
 67,
 69,
 68,
 69,
 68,
 68,
 68,
 69,
 69,
 68,
 67,
 67,
 67,
 66,
 69,
 70,
 69,
 69,
 70,
 67,
 67,
 69,
 67,
 67,
 67,
 69,
 71,
 70,
 70,
 69,
 70,
 68,
 68,
 69,


In [19]:
# check resting heart rate
print("number of empty values for resting heart rate: ", np.isnan(restingHR).sum())
print("list length of resting heart rate: ", len(restingHR))

number of empty values for resting heart rate:  0
list length of resting heart rate:  261


`restingHR` has sucessfully been filled with resting heart rate data. It consists of 261 data points with no Nan values.

In [20]:
# Creating dataframe
restingHRdf = pd.DataFrame(restingHR,index = dates) # use date as row index
restingHRdf.columns = ['resting_HR'] # rename columns

# Check
restingHRdf

resting_HR
2021-02-25          66
2021-02-26          67
2021-02-27          66
2021-02-28          66
2021-03-01          65
...                ...
2021-11-08          67
2021-11-09          66
2021-11-10          65
2021-11-11          63
2021-11-12          64

[261 rows x 1 columns]

Resting heart rate data has been successfully converted into a dataframe under the name `restingHRdf`. The data is present under the column name `resting_HR`.

##### Exporting `dailyHRdf` into csv

In [181]:
restingHRdf.to_csv(r"C:\Users\gsagararatne\Documents\BrainStation files\portfolio projects\FitbitSleepAnalysis\restingHR_data.csv", index=False, header=True)

#### Sleep data
#####  Exploring sleep data

In [11]:
# Loading in a random date
sleep = open('sleep2021-02-25.json')

# Loading json
sleepdata = json.load(sleep)
sleepdata = json.loads(sleepdata)

# Making it easier to read
print(json.dumps(sleepdata, indent=4, sort_keys=True))

{
    "sleep": [
        {
            "dateOfSleep": "2021-02-25",
            "duration": 22620000,
            "efficiency": 83,
            "endTime": "2021-02-25T09:22:30.000",
            "infoCode": 0,
            "isMainSleep": true,
            "levels": {
                "data": [
                    {
                        "dateTime": "2021-02-25T03:05:00.000",
                        "level": "wake",
                        "seconds": 540
                    },
                    {
                        "dateTime": "2021-02-25T03:14:00.000",
                        "level": "light",
                        "seconds": 240
                    },
                    {
                        "dateTime": "2021-02-25T03:18:00.000",
                        "level": "wake",
                        "seconds": 1170
                    },
                    {
                        "dateTime": "2021-02-25T03:37:30.000",
                        "level": "light",
               

In [79]:
# Sleep dictionary keys
sleepdata.keys()

dict_keys(['sleep', 'summary'])

There seems to be 2 different keys: `sleep` and `summary`.

In [3]:
# Sleep key
sleepdata["sleep"]

[{'dateOfSleep': '2021-02-25',
  'duration': 22620000,
  'efficiency': 83,
  'endTime': '2021-02-25T09:22:30.000',
  'infoCode': 0,
  'isMainSleep': True,
  'levels': {'data': [{'dateTime': '2021-02-25T03:05:00.000',
     'level': 'wake',
     'seconds': 540},
    {'dateTime': '2021-02-25T03:14:00.000', 'level': 'light', 'seconds': 240},
    {'dateTime': '2021-02-25T03:18:00.000', 'level': 'wake', 'seconds': 1170},
    {'dateTime': '2021-02-25T03:37:30.000', 'level': 'light', 'seconds': 1170},
    {'dateTime': '2021-02-25T03:57:00.000', 'level': 'deep', 'seconds': 720},
    {'dateTime': '2021-02-25T04:09:00.000', 'level': 'light', 'seconds': 1260},
    {'dateTime': '2021-02-25T04:30:00.000', 'level': 'deep', 'seconds': 300},
    {'dateTime': '2021-02-25T04:35:00.000', 'level': 'light', 'seconds': 420},
    {'dateTime': '2021-02-25T04:42:00.000', 'level': 'deep', 'seconds': 390},
    {'dateTime': '2021-02-25T04:48:30.000', 'level': 'light', 'seconds': 150},
    {'dateTime': '2021-02-25T

In [9]:
# Summary key
sleepdata["summary"]

{'stages': {'deep': 59, 'light': 208, 'rem': 34, 'wake': 76},
 'totalMinutesAsleep': 301,
 'totalSleepRecords': 1,
 'totalTimeInBed': 377}

In the key `sleep`, the items that interest me are:
- efficiency

In the key `summary`, the items that interest me are:
- stages
    - deep
    - light
    - rem
    - wake
- totalMinutesAsleep
- totalTimeInBed

These are the items I plan to extract and analyse down the line.

In [5]:
# Checking format of extraction

print("sleep accuracy: ", sleepdata["sleep"][0]["efficiency"])

print("deep sleep: ", sleepdata["summary"]["stages"]["deep"])
print("light sleep: ", sleepdata["summary"]["stages"]["light"])
print("rem sleep: ", sleepdata["summary"]["stages"]["rem"])
print("wake time: ", sleepdata["summary"]["stages"]["wake"])
print("total mins asleep: ", sleepdata["summary"]["totalMinutesAsleep"])
print("total time in bed: ", sleepdata["summary"]["totalTimeInBed"])

sleep accuracy:  83
deep sleep:  59
light sleep:  208
rem sleep:  34
wake time:  76
total mins asleep:  301
total time in bed:  377


##### Extracting sleep data

In [101]:
# Lists to store respective data
sleep_efficiency = []
deep_sleep = []
light_sleep = []
rem_sleep = []
wake_sleep = []
total_mins_asleep = []
total_time_in_bed = []

dates = pd.date_range('2021-02-25', '2021-11-12')

for date in dates:
    fname = 'sleep'+date.strftime('%Y-%m-%d')+'.json'
    with open(fname) as f:
        date_data_sleep = json.loads(f.read())
        datasleep = pd.read_json(date_data_sleep, typ='series')
        
        # sleep efficiency
        try:
            sleep_efficiency.append(datasleep["sleep"][0]["efficiency"])
        except:
            sleep_efficiency.append(np.nan)
        
        # deep sleep
        try:
            deep_sleep.append(datasleep["summary"]["stages"]["deep"])
        except:
            deep_sleep.append(np.nan)
           
            
        # light sleep
        try:
            light_sleep.append(datasleep["summary"]["stages"]["light"])
        except:
            light_sleep.append(np.nan)
            
            
        # rem sleep
        try:
            rem_sleep.append(datasleep["summary"]["stages"]["rem"])
        except:
            rem_sleep.append(np.nan)           
            
        # wake sleep
        try:
            wake_sleep.append(datasleep["summary"]["stages"]["wake"])
        except:
            wake_sleep.append(np.nan)                   

        # mins asleep    
        try:
            total_mins_asleep.append(datasleep['summary']['totalMinutesAsleep'])
        except:
            total_mins_asleep.append(np.nan)            
            
        # time in bed  
        try:
            total_time_in_bed.append(datasleep['summary']['totalTimeInBed'])
        except:
            total_time_in_bed.append(np.nan)

In [85]:
# check for nan values
print("number of empty values for sleep efficiency: ", np.isnan(sleep_efficiency).sum())
print("number of empty values for deep sleep: ", np.isnan(deep_sleep).sum())
print("number of empty values for light sleep: ", np.isnan(light_sleep).sum())
print("number of empty values for rem sleep: ", np.isnan(rem_sleep).sum())
print("number of empty values for wake sleep: ", np.isnan(wake_sleep).sum())
print("number of empty values for total mins asleep: ", np.isnan(total_mins_asleep).sum())
print("number of empty values for total time in bed: ", np.isnan(total_time_in_bed).sum())

number of empty values for sleep efficiency:  7
number of empty values for deep sleep:  7
number of empty values for light sleep:  7
number of empty values for rem sleep:  7
number of empty values for wake sleep:  7
number of empty values for total mins asleep:  0
number of empty values for total time in bed:  0


In [86]:
print("list length of sleep efficiency: ", len(sleep_efficiency))

print("list length of deep sleep: ", len(deep_sleep))

print("list length of light sleep: ", len(light_sleep))

print("list length of rem sleep: ", len(rem_sleep))

print("list length of wake sleep: ", len(wake_sleep))

print("list length of total mins asleep: ", len(total_mins_asleep))

print("list length of total time in bed: ", len(total_time_in_bed))

list length of sleep efficiency:  261
list length of deep sleep:  261
list length of light sleep:  261
list length of rem sleep:  261
list length of wake sleep:  261
list length of total mins asleep:  261
list length of total time in bed:  261


My sleep data consists of 261 rows of data. I had to implement **try** and **except** statements in my extracting loop because `stages` was absent under `summary` in the `datasleep` dictionary for 7 json files. This is because I forgot to wear my fitbit to sleep. Hence, had to replace said missing values with **Nans**.

In [87]:
# Creating the sleepdf dataframe
sleepdf = pd.DataFrame([sleep_efficiency, deep_sleep, light_sleep, rem_sleep, wake_sleep, total_mins_asleep, total_time_in_bed]) 
sleepdf = sleepdf.transpose()
sleepdf.columns = ["sleep_efficiency", "deep_sleep", "light_sleep", "rem_sleep", "wake_sleep", "total_mins_asleep", "total_time_in_bed"]
sleepdf = sleepdf.set_index(dates)
print(sleepdf.shape)

# Check
sleepdf.head()

(261, 7)


sleep_efficiency  deep_sleep  light_sleep  rem_sleep  wake_sleep  \
2021-02-25              83.0        59.0        208.0       34.0        76.0   
2021-02-26              89.0        79.0        221.0       80.0        63.0   
2021-02-27              92.0        65.0        202.0       47.0        61.0   
2021-02-28              93.0        91.0        191.0       54.0        38.0   
2021-03-01              87.0        89.0        102.0       44.0        59.0   

            total_mins_asleep  total_time_in_bed  
2021-02-25              312.0              377.0  
2021-02-26              393.0              443.0  
2021-02-27              344.0              375.0  
2021-02-28              346.0              374.0  
2021-03-01              257.0              294.0

##### Exporting `sleepdf` into csv

In [136]:
sleepdf.to_csv(r"C:\Users\gsagararatne\Documents\BrainStation files\portfolio projects\FitbitSleepAnalysis\sleep_data.csv", index=False, header=True)

#### Step data
##### Exploring step data

**Note:** As I was randomly loading a few step json files, I noticed that it automatically loads step data from the opened date (start) till 2021-11-13, which happens to be 1 day past the extracted end date (2021-11-12).

In [97]:
# Loading in a first date
step = open('step2021-02-25.json')

# Loading json
stepdata = json.load(step)
stepdata = json.loads(stepdata)

# Making it easier to read
print(json.dumps(stepdata, indent=4, sort_keys=True))

{
    "activities-steps": [
        {
            "dateTime": "2021-02-25",
            "value": "7248"
        },
        {
            "dateTime": "2021-02-26",
            "value": "5496"
        },
        {
            "dateTime": "2021-02-27",
            "value": "6568"
        },
        {
            "dateTime": "2021-02-28",
            "value": "5965"
        },
        {
            "dateTime": "2021-03-01",
            "value": "2990"
        },
        {
            "dateTime": "2021-03-02",
            "value": "3563"
        },
        {
            "dateTime": "2021-03-03",
            "value": "7981"
        },
        {
            "dateTime": "2021-03-04",
            "value": "6383"
        },
        {
            "dateTime": "2021-03-05",
            "value": "8235"
        },
        {
            "dateTime": "2021-03-06",
            "value": "6339"
        },
        {
            "dateTime": "2021-03-07",
            "value": "6118"
        },
        {
     

In [89]:
stepdata.keys()

dict_keys(['activities-steps'])

In [90]:
len(stepdata["activities-steps"])

262

`stepdata` seems to only have one key called `activities-steps` and has 262 items. My plan is to loop through the step data as I did with the sleep data and see if I'm able to extract only the 261 days worth of data.

##### Extracting step data

In [102]:
dates = pd.date_range('2021-02-25', '2021-11-12')

steps = []
for date in dates:
    fname = 'step' + date.strftime('%Y-%m-%d') + '.json'
    with open(fname) as f:
        date_data = json.loads(f.read())
        stepsdata = pd.read_json(date_data, typ='series')
        steps.append(stepsdata['activities-steps'][0]) # the zero indexes the first value

stepsdf = pd.DataFrame(steppers,index = dates) # use date as row index
stepsdf.columns = ['date','steps'] # rename columns
stepsdf.drop("date",axis=1, inplace=True)

In [108]:
stepsdf.head(3)

steps
2021-02-25  7248
2021-02-26  5496
2021-02-27  6568

In [107]:
stepsdf.tail(3)

steps
2021-11-10  10023
2021-11-11   9524
2021-11-12  15507

In [105]:
stepsdf.shape

(261, 1)

In [114]:
stepsdf.isna().sum()

steps    0
dtype: int64

Can confirm that my step data from **2021-02-25** to **2021-11-12** has been successfully extracted with 0 missing values. 

##### Exporting `stepsdf` into csv

In [137]:
stepsdf.to_csv(r"C:\Users\gsagararatne\Documents\BrainStation files\portfolio projects\FitbitSleepAnalysis\steps_data.csv", index=False, header=True)

#### Activity data

##### Exploring activity data

In [128]:
# Loading in a random date
act = open('act2021-02-25.json')

# Loading json
actdata = json.load(act)
actdata = json.loads(actdata)

# Making it easier to read
print(json.dumps(actdata, indent=4, sort_keys=True))

{
    "activities": [],
    "goals": {
        "activeMinutes": 30,
        "caloriesOut": 2563,
        "distance": 8.05,
        "steps": 10000
    },
    "summary": {
        "activeScore": -1,
        "activityCalories": 1084,
        "caloriesBMR": 1490,
        "caloriesOut": 2411,
        "distances": [
            {
                "activity": "total",
                "distance": 5.2
            },
            {
                "activity": "tracker",
                "distance": 5.2
            },
            {
                "activity": "loggedActivities",
                "distance": 0
            },
            {
                "activity": "veryActive",
                "distance": 1.03
            },
            {
                "activity": "moderatelyActive",
                "distance": 0.48
            },
            {
                "activity": "lightlyActive",
                "distance": 3.67
            },
            {
                "activity": "sedentaryActive",
 

In [129]:
actdata.keys()

dict_keys(['activities', 'goals', 'summary'])

Seems like actdata has 3 keys: activities, goals and summary.
In the key `summary`, the items that interest me are:
- caloriesOut
- distances
    - "activity":"total"
   
I did a bit of crosschecking with my fitbit app, and the calories burnt in a day is listed under `caloriesOut`. On top of that, the total distance walked in a day (km) is listed under total activity. These are the items I plan to extract and analyse down the line.

In [138]:
# format to extract caloriesOut
actdata["summary"]["caloriesOut"]

2411

In [141]:
# format to extract total activity
actdata["summary"]["distances"][0]["distance"]

5.2

##### Extacting activity data

In [143]:
# Lists to store respective data
calories_burnt = []
distance_km = []


# dates = pd.date_range('2021-02-25', '2021-02-26')
dates = pd.date_range('2021-02-25', '2021-11-12')

for date in dates:
    fname = 'act'+date.strftime('%Y-%m-%d')+'.json'
    with open(fname) as f:
        date_data_act = json.loads(f.read())
        dataact = pd.read_json(date_data_act, typ='series')
            
        calories_burnt.append(dataact["summary"]["caloriesOut"])   
        distance_km.append(dataact["summary"]["distances"][0]["distance"])

In [144]:
# Check
calories_burnt

[2411,
 2203,
 2302,
 2285,
 1994,
 2085,
 2446,
 2265,
 2346,
 2227,
 2309,
 2183,
 1971,
 2255,
 2004,
 2640,
 2130,
 2632,
 2088,
 1964,
 2251,
 2012,
 2575,
 2218,
 2832,
 2119,
 2355,
 2113,
 2290,
 2134,
 2145,
 2212,
 2309,
 3020,
 2195,
 2199,
 2240,
 2849,
 2315,
 2122,
 2278,
 2247,
 2018,
 2173,
 2611,
 2232,
 2350,
 2200,
 2233,
 2311,
 2092,
 2026,
 2348,
 2315,
 2274,
 2200,
 2830,
 2677,
 2432,
 2796,
 3435,
 2339,
 2090,
 2145,
 2259,
 2421,
 2085,
 2181,
 2191,
 2150,
 2166,
 2153,
 2415,
 2259,
 2200,
 2268,
 2184,
 2284,
 2276,
 2491,
 2453,
 2213,
 2354,
 2220,
 2063,
 2263,
 2277,
 2416,
 2176,
 2221,
 2025,
 2053,
 1994,
 2333,
 2449,
 2286,
 2111,
 2339,
 2417,
 2501,
 2433,
 2459,
 2519,
 2446,
 2345,
 2557,
 2441,
 2592,
 2365,
 2459,
 2563,
 2217,
 2288,
 2458,
 2265,
 2273,
 2609,
 2514,
 2347,
 2414,
 2231,
 2360,
 2291,
 2212,
 2184,
 2335,
 2432,
 2204,
 2238,
 1736,
 2129,
 2109,
 1990,
 1990,
 2175,
 2182,
 2143,
 2206,
 2075,
 1995,
 1958,
 2138,
 2077,

In [145]:
# Check
distance_km

[5.2,
 3.7,
 4.43,
 4.15,
 2.02,
 2.4,
 5.51,
 4.3,
 5.55,
 4.27,
 4.12,
 2.5,
 1.5,
 5,
 2.56,
 9.14,
 4.39,
 9.1,
 2.84,
 2.2,
 4.21,
 2.49,
 6.77,
 4.16,
 12.69,
 3.55,
 0.275,
 4.12,
 4.27,
 3.63,
 3.58,
 4.3,
 5.01,
 13.33,
 3.76,
 2.95,
 3.2800000000000002,
 5.38,
 4.42,
 2.5300000000000002,
 3.48,
 3.9699999999999998,
 2.06,
 3.5,
 7.21,
 3.16,
 5.13,
 4.41,
 3.91,
 4.93,
 3.21,
 3.27,
 4.4,
 4.55,
 4.4,
 4.26,
 10.6,
 8.85,
 5.58,
 9.53,
 13.38,
 4.74,
 3.36,
 3.24,
 3.22,
 5.38,
 3.19,
 3.75,
 4.34,
 3.84,
 3.38,
 3.38,
 5.82,
 4.79,
 4.36,
 4.75,
 3.95,
 4.89,
 4.72,
 7.44,
 5.52,
 4.2,
 5.26,
 3.98,
 2.52,
 4.83,
 4,
 5.42,
 4.69,
 5.23,
 3,
 2.84,
 2.52,
 5.12,
 6.73,
 5.16,
 3.5,
 5.69,
 5.99,
 6.84,
 5.22,
 6.97,
 7.1,
 7.11,
 7.02,
 8.92,
 6.22,
 7.5600000000000005,
 6.19,
 7.64,
 7.99,
 5.42,
 6.05,
 7.18,
 5.88,
 5.63,
 8.87,
 0.375,
 5.62,
 6.19,
 3.5300000000000002,
 4.37,
 5.79,
 4.24,
 4.78,
 4.9,
 6.29,
 3.67,
 4.85,
 1.31,
 3.19,
 2.84,
 2.05,
 2.66,
 3.33,
 4.35

In [147]:
# check calories_burnt
print("number of empty values for calories burnt: ", np.isnan(calories_burnt).sum())
print("list length of calories burnt: ", len(calories_burnt))

# check distance_km
print("number of empty values for distance in km: ", np.isnan(distance_km).sum())
print("list length of distance in km: ", len(distance_km))

number of empty values for calories burnt:  0
list length of calories burnt:  261
number of empty values for distance in km:  0
list length of distance in km:  261


`calories_burnt` and `distance_km` has been successfully filled with calories burnt and distance walked in a day (km) respectively. They both consist of 261 data points with no Nan values.

In [148]:
# Creating a activity dataframe
actdf = pd.DataFrame([calories_burnt, distance_km]) 
actdf = actdf.transpose()
actdf.columns = ["calories_burnt","distance_km"]
actdf = actdf.set_index(dates)
print(actdf.shape)
actdf.head()

(261, 2)


calories_burnt  distance_km
2021-02-25          2411.0         5.20
2021-02-26          2203.0         3.70
2021-02-27          2302.0         4.43
2021-02-28          2285.0         4.15
2021-03-01          1994.0         2.02

##### Exporting `actdf` data

In [149]:
actdf.to_csv(r"C:\Users\gsagararatne\Documents\BrainStation files\portfolio projects\FitbitSleepAnalysis\act_data.csv", index=False, header=True)

### Summary
- Great opportunity to understand how fitbit's API works and to get comfortable working with json data. 
- 261 days' worth of my very own data has been collected. The particular items I've extracted and choose to analyse further are:
    - sleep_efficiency 
    - deep_sleep 
    - light_sleep 
    - rem_sleep 
    - wake_sleep 
    - total_mins_asleep 
    - total_time_in_bed
    - steps
    - calories_burnt
    - distance_km
    
- I am aware that 7 rows of sleep data is missing and that's mainly because I've either forgotten to wear it to bed or the fitbit was loose on my wrist.
- 3 csv files have been exported from this notebook. I plan to create a new notebook where I further explore my data under one whole dataframe. 